Implement `omb theme`

RobLoach opened the pull request:
This introduces a `themes` plugin which allows changing the theme on the fly. This mimics the [themes oh-my-zsh plugin](https://github.com/ohmyzsh/ohmyzsh/tree/master/plugins/themes).

## Usage

`theme <theme_name>` - Changes the theme to specified theme.

`theme ` - Changes the theme to some random theme.

`lstheme ` - Lists installed themes.

RobLoach committed:
plugins: Add Theme plugin

akinomyoga commented:
Thank you for the PR! I have a question. How is the upstream OMZ `themes` plugin different from `omz theme` [1]?

- [1] https://github.com/ohmyzsh/ohmyzsh/blob/f4423ebd09fbc7670815c3c20cc86c81b7319e5f/lib/cli.zsh#L658-L796

Since this is related to the OMB specific feature for themes, I feel this should be implemented as a subcommand of the `omb` command similar to `omz theme`. However, the upstream OMZ seems to contain quite similar features in two locations, `plugins/themes` and `omz theme`, which appears to be weird to me.

RobLoach commented:
Namespacing it within `omb` makes so much sense. Thanks for linking the stub.

RobLoach committed:
Move to Usage: omb theme <command>

Available commands:
  list          List all available themes
  set <theme>   Set the given theme in your .bashrc file
  use <theme>   Load the given theme

RobLoach commented:
Not sure how I feel about manipulating the `.bashrc` file with `omb theme set zork`.

akinomyoga commented on lib/cli.bash:
```diff
@@ -18,7 +18,43 @@ function _omb_cmd_reload {
+  list)
+      local -a available_themes
+      _comp_cmd_omb__get_available_themes
+      for theme in "${available_themes[@]}"; do
+        echo "$theme"
+      done
```
```suggestion
      _omb_util_print_lines "${available_themes[@]}"
```

akinomyoga commented on lib/cli.bash:
```diff
@@ -18,7 +18,43 @@ function _omb_cmd_reload {
 }
 function _omb_cmd_theme {
-  echo 'Not yet implemented'
+  case "$1" in
```
In case the user sets `set -u`:

```suggestion
  case "${1-}" in
```

akinomyoga commented on lib/cli.bash:
```diff
@@ -18,7 +18,43 @@ function _omb_cmd_reload {
+      done
+    ;;
+  use)
+    local theme=$2
```
```suggestion
    local theme=${2-}
```

akinomyoga commented on lib/cli.bash:
```diff
@@ -18,7 +18,43 @@ function _omb_cmd_reload {
+    _comp_cmd_omb__get_available_themes
+    for i in "${available_themes[@]}"; do
+      if [ "$i" == "$theme" ]; then
+        _omb_module_require_theme "$theme"
```
`_omb_module_require_theme` is kind of "include once"; it checks whether the module is already loaded and loads the module only when it hasn't been loaded. So, with this implementation, one cannot switch to the theme that was loaded before.

To handle this, we may need to factorize a part of `_omb_module_require` (`oh-my-bash.sh`) into `_omb_module_search` and use it to obtain the location of the theme file.

akinomyoga commented on lib/cli.bash:
```diff
@@ -18,7 +18,43 @@ function _omb_cmd_reload {
+        return 0
+      fi
+    done
+    echo "Theme '$theme' not found"
```
```suggestion
    _omb_util_print "Theme '$theme' not found"
```

We avoid using `echo` because `echo` is affected by the setting `shopt -s xpg_echo` and is not trustable.

akinomyoga commented on lib/cli.bash:
```diff
@@ -18,7 +18,43 @@ function _omb_cmd_reload {
+    echo 'Not yet implemented'
+    ;;
+  *)
+    echo 'Usage: omb theme <command>'
+    echo ''
+    echo 'Available commands:'
+    echo '  list          List all available themes'
+    echo '  set <theme>   Set the given theme in your .bashrc file'
+    echo '  use <theme>   Load the given theme'
```
```suggestion
    _omb_util_print_lines \
      'Usage: omb theme <command>' \
      '' \
      'Available commands:' \
      '  list          List all available themes' \
      '  set <theme>   Set the given theme in your .bashrc file' \
      '  use <theme>   Load the given theme'
```

akinomyoga commented on lib/cli.bash:
```diff
@@ -18,7 +18,43 @@ function _omb_cmd_reload {
+  use)
+    local theme=$2
+    if [[ -z "$theme" ]]; then
+      echo 'Usage: omb theme use <theme>'
```
```suggestion
      _omb_util_print 'Usage: omb theme use <theme>' >&2
```

Error messages should be output to fd 2 (`stderr`).

RobLoach committed:
Update lib/cli.bash

Co-authored-by: Koichi Murase <myoga.murase@gmail.com>

RobLoach committed:
Apply suggestions from code review

Co-authored-by: Koichi Murase <myoga.murase@gmail.com>

RobLoach committed:
Merge branch 'master' of https://github.com/ohmybash/oh-my-bash into plugin-themes

RobLoach committed:
Return 1 when the theme isn't found

RobLoach committed:
Use available_themes function
